### PR TITLE
Add Qwix quantization support

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

In this PR, I do the following:
* Add support for weight/activation quantization through Qwix
* Enable a new `additional_config.quantization` path that accepts a quantization file name (pointing to files in `tpu_commons/models/jax/utils/quantization/configs/`) to define the Qwix config to use

The format of this file is currently:

```
qwix:
  rules:
    # NOTE: each entry corresponds to a qwix.QuantizationRule
    - module_path: '.*attn.*'
      weight_qtype: 'int8'
    - module_path: '.*'
      weight_qtype: 'int8'
      act_qtype: 'int8'
    ....

```

* Update the README with quantization details
* Add pre-model loading config checks to `tpu_commons/platforms/tpu_jax.py`
* Create a helper function for creating KV caches (since the logic is duplicated now)

# Verification
## Memory usage decrease
16.43 / chip -> 8.56 / chip
```
INFO 07-21 21:19:17 [utils.py:236] Init model before quant hbm=[(16.43, 31.25), (16.43, 31.25), (16.43, 31.25), (16.43, 31.25), (16.43, 31.25), (16.43, 31.25), (16.43, 31.25), (16.43, 31.25)]Gb


INFO 07-21 21:19:42 [tpu_jax_runner.py:185] Init model after quant | hbm=[(8.56, 31.25), (8.56, 31.25), (8.56, 31.25), (8.56, 31.25), (8.56, 31.25), (8.56, 31.25), (8.56, 31.25), (8.56, 31.25)]Gb
```

## Accuracy and throughput
We know that Llama3-70B has [performance issues with quantization], so a good sanity check is to ensure that the relative drop between MaxText (AQT) and TPU Commons (Qwix) is similar.  I give more details in the relevant [bug](https://b.corp.google.com/issues/425362407#comment18) but here's a quick summary:

### MMLU (5000 requests)

| **Framework**      | **Quantization** | **Throughput (tokens/s)** | **Accuracy on MMLU (5000 requests)** |
|--------------------|------------------|---------------------------|--------------------------------------|
| MaxText (AQT)      | N/A              | 3275.68                   | 0.7932                               |
| TPU Commons (Qwix) | N/A              | 19677.90                  | 0.8226                               |
| MaxText (AQT)      | int8 --w8a8      | 4129.40                   | 0.5278                               |
| TPU Commons (Qwix) | int8 --w8a8      | 20626.22                  | 0.4928                               |

### MLPerf (6000 requests)

| **Framework**      | **Quantization** | **Throughput (tokens/s)** | **ROUGE Result on MLPerf (6000 requests)**                                                       |
|--------------------|------------------|---------------------------|--------------------------------------------------------------------------------------------------|
| MaxText (AQT)      | N/A              | 3289.61                   | {'rouge1': 41.5671, 'rouge2': 18.2776, 'rougeL': 26.1648, 'rougeLsum': 39.3219, 'gen_num': 6000} |
| TPU Commons (Qwix) | N/A              | 4016.5                    | {'rouge1': 45.4742, 'rouge2': 21.4921, 'rougeL': 28.5965, 'rougeLsum': 43.1276, 'gen_num': 6000} |
| MaxText (AQT)      | int8 --w8a8      | 4259.92                   | {'rouge1': 29.084, 'rouge2': 13.5394, 'rougeL': 20.7492, 'rougeLsum': 27.1464, 'gen_num': 6000}  |
| TPU Commons (Qwix) | int8 --w8a8      | 4070.12^                  | {'rouge1': 33.9784, 'rouge2': 17.0402, 'rougeL': 24.7212, 'rougeLsum': 31.847, 'gen_num': 6000}" |

^ I am investigating why we aren't seeing a throughput gain [here](https://b.corp.google.com/issues/433350117)

# Tests

Added relevant testing in `tests/` and also added a new E2E test in `.buildkite/pipeline_jax.yml`.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -224,3 +224,30 @@ While this will run the code in a Docker image, you can also run the bare `tests
...
+To enable quantization, you can specify a quantization config filename found inside the quantization config directory (`tpu_commons/models/jax/utils/quantization/configs/`), for example:
+
+```
+... --additional_config='{"quantization": "int8_default.yaml"}'
+```
+
+### Creating your own quantization config
+To create your own quantization:
+
+1. Add a new file to the quantization config irectory (`tpu_commons/models/jax/utils/quantization/configs/`)
```
Comment (gpolovets1) [inline README.md]: 'directory'

File: tests/e2e/benchmarking/mmlu.sh [hunk 2]
```diff
@@ -32,6 +32,11 @@ if [ "$NEW_MODEL_DESIGN" = "True" ]; then
     model_list="meta-llama/Llama-3.1-8B-Instruct"
     extra_serve_args+=("--hf_overrides")
     extra_serve_args+=('{"architectures": ["Llama3_8B"]}')
+    if [ "$QUANTIZATION" = "True" ]; then
+        echo "QUANTIZATION is True. Running with quantization."
+        extra_serve_args+=("--additional_args")
+        extra_serve_args+=('{"quantization": {"rules_file": "tpu_commons/models/jax/utils/quantization/quantize_all_modules_int8_wa.yaml"}}')
```
Comment (mitalisi) [inline tests/e2e/benchmarking/mmlu.sh]: this needs to be updated. {"quantization": "quantize_all_modules_int8_wa.yaml"}

File: tpu_commons/models/jax/model_loader.py [hunk 3]
```diff
@@ -39,6 +41,56 @@ def _get_model_architecture(config: PretrainedConfig) -> nnx.Module:
         f"Supported architectures: {list(_MODEL_REGISTRY.keys())}")
 
 
+def _maybe_apply_qwix_quantization(vllm_config: VllmConfig, model: nnx.Module,
+                                   rng: jax.Array, mesh: Mesh) -> nnx.Module:
+    """
+    Will apply Qwix quantization if either dtype or rules file is provided.  If dtype is provided,
```
Comment (mitalisi) [inline tpu_commons/models/jax/model_loader.py]: Update comment "if quantization file is configured". We should also check in parsing config commands for missing file if not doing so already.

File: tpu_commons/models/jax/utils/quantization/quantization_utils.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    )
+
+    def _device_array(*args, sharding=None, **kwargs) -> jax.Array:
+        if sharding is None:
+            sharding = NamedSharding(mesh, PartitionSpec(None))
+        return jax.device_put(*args, device=sharding, **kwargs)
+
+    input_ids = jax.random.randint(rng, (512, ), 0, 100, dtype=jnp.int32)
+    positions = jax.random.randint(rng, (512, ), 0, 100, dtype=jnp.int32)
+    slot_mapping_metadata = jax.random.randint(rng, (3, 512),
```
Comment (gpolovets1) [inline tpu_commons/models/jax/utils/quantization/quantization_utils.py]: Where are these hardcodings coming from? Doesn't it depend on the settings used in the model/vllm config?

File: tpu_commons/models/jax/utils/quantization/quantization_utils.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                                               0,
+                                               100,
+                                               dtype=jnp.int32)
+    num_slices = jax.random.randint(rng, (1, ), 0, 100, dtype=jnp.int32)
+    block_tables = jax.random.randint(rng, (256, 16), 0, 100, dtype=jnp.int32)
```
Comment (gpolovets1) [inline tpu_commons/models/jax/utils/quantization/quantization_utils.py]: 16 = kv_cache_block_size?

File: tpu_commons/models/jax/model_loader.py [hunk 6]
```diff
@@ -39,6 +41,56 @@ def _get_model_architecture(config: PretrainedConfig) -> nnx.Module:
...
+        mesh: the mesh to use
+
+    Returns:
+        the potentially quantized model
+    """
+    # NOTE: we expect the value of "quantization" to be the name of a file in `tpu_commons/models/jax/utils/quantization/configs`
+    # if given
+    maybe_qwix_config = None
+    if vllm_config.additional_config.get("quantization"):
+        maybe_quantization_config = convert_quantization_config_file_path_to_dict(
```
Comment (gpolovets1) [inline tpu_commons/models/jax/model_loader.py]: nit: can remove 'convert_' for a bit more succinctness.
Also, maybe_ I think is typically referred to methods. The return itself is a quantization_config but it just so happens that it can also be null.

File: tests/e2e/benchmarking/mmlu.sh [hunk 7]
```diff
@@ -32,6 +32,11 @@ if [ "$NEW_MODEL_DESIGN" = "True" ]; then
     model_list="meta-llama/Llama-3.1-8B-Instruct"
     extra_serve_args+=("--hf_overrides")
     extra_serve_args+=('{"architectures": ["Llama3_8B"]}')
+    if [ "$QUANTIZATION" = "True" ]; then
```
Comment (gpolovets1) [inline tests/e2e/benchmarking/mmlu.sh]: Do we need this as an env variable? Wondering why we don't just hardcode this for specific e2e tests.

File: tpu_commons/models/jax/utils/quantization/quantization_utils.py (hunk 4 above)
Comment (jrplatin) [reply tpu_commons/models/jax/utils/quantization/quantization_utils.py]: Good question -- I just took these values from a given forward pass for Llama3-8.1B but I am not sure if the actual values matter that much.  @liudangyi do you happen to know whether the `model_inputs` here matter for `qwix.quantize_model`?

Comment (jrplatin) [reply tpu_commons/models/jax/utils/quantization/quantization_utils.py]: But also good point, let me make them constants and the naming more clear so it's not just random values

File: tests/e2e/benchmarking/mmlu.sh (hunk 7 above)
Comment (jrplatin) [reply tests/e2e/benchmarking/mmlu.sh]: How would you hardcode this without duplicating the file?

File: tests/e2e/benchmarking/mmlu.sh (hunk 2 above)
Comment (jrplatin) [reply tests/e2e/benchmarking/mmlu.sh]: Good catch, thanks!

File: tpu_commons/models/jax/utils/quantization/quantization_utils.py (hunk 4 above)
Comment (liudangyi) [reply tpu_commons/models/jax/utils/quantization/quantization_utils.py]: The `model_inputs` are used to run the model once so that Qwix knows how to quantize each weights. The inputs don't need to match the actual ones, as long as the consumed weights are the same. The actual numerics of the inputs don't matter so it's fine to use all zeros. If your model supports running without a KV cache, it's perfectly fine to skip the KV cache too.

File: tpu_commons/models/jax/model_loader.py [hunk 8]
```diff
@@ -39,6 +41,56 @@ def _get_model_architecture(config: PretrainedConfig) -> nnx.Module:
...
+        model = nnx.jit(qwix_quantize_nnx_model_with_config,
+                        donate_argnums=(0, ),
+                        static_argnames=(
+                            "mesh",
+                            "num_hidden_layers",
+                            "kv_cache_block_size",
+                            "kv_cache_num_kv_heads",
+                            "kv_cache_head_size",
+                        ))(model=model,
+                           rng=rng,
```
Comment (liudangyi) [inline tpu_commons/models/jax/model_loader.py:57]: Since you already use `functools.partial`, it's more concise to put all the static arguments in the partial object, e.g. 

```py
model = nnx.jit(
    functools.partial(
        qwix_quantize_nnx_model,
        qwix_config=maybe_qwix_config,
        rng=rng,
        mesh=mesh,
        num_hidden_layers=vllm_config.model_config.hf_config.num_hidden_layers,
        kv_cache_block_size=block_size,
        kv_cache_num_kv_heads=num_kv_heads,
        kv_cache_head_size=head_size,
    ),
    donate_argnums=(0,),
)(model)
```

File: tpu_commons/models/jax/utils/quantization/quantization_utils.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+from tpu_commons import utils_jax as utils
+from tpu_commons.logger import init_logger
+from tpu_commons.models.jax.attention_metadata import AttentionMetadata
+from tpu_commons.runner.utils import create_kv_caches
+
+logger = init_logger(__name__)
+
+MAX_INT8 = 127.5
+MAX_INT4 = 7.5
+E4M3_MAX = jnp.finfo(jnp.float8_e4m3fn).max.astype(jnp.float32)
```
Comment (liudangyi) [inline tpu_commons/models/jax/utils/quantization/quantization_utils.py]: Are these constants used? They are the implementation details of Qwix and shouldn't be assumed.

File: tpu_commons/runner/utils.py [hunk 10]
```diff
@@ -123,3 +129,85 @@ def __exit__(self, exc_type, exc_value, traceback):
             pxla._cached_lowering_to_hlo = self._original_func
         # Don't suppress any exceptions that occurred inside the 'with' block
         return False
+
+
+def get_jnp_dtype_from_str(dtype_str: str) -> type:
```
Comment (liudangyi) [inline tpu_commons/runner/utils.py]: This is usually redundant because you can pass a str to most dtype argument directly, e.g. `jnp.zeros((16, 16), dtype="float8_e4m3fn")` just works.

File: tpu_commons/runner/utils.py [hunk 11]
```diff
@@ -123,3 +129,85 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+                f"dtype={cache_dtype} | "
+                f"hbm={utils.hbm_usage_gb(devices)}Gb")
+
+    def _allocate() -> jax.Array:
+        return jnp.empty(
+            shape=cache_shape,
+            dtype=cache_dtype,
+        )
+
+    sharded_allocate = jax.jit(_allocate, out_shardings=sharding)
```
Comment (liudangyi) [inline tpu_commons/runner/utils.py:242]: nit: you don't have to use a jax.jit to specify a sharding. Instead, use the `out_sharding` argument in e.g. `jnp.empty`.

File: tpu_commons/models/jax/utils/quantization/quantization_utils.py (hunk 9 above)
Comment (jrplatin) [reply tpu_commons/models/jax/utils/quantization/quantization_utils.py]: Ah good catch, that was KV cache quant thanks!

File: tpu_commons/models/jax/model_loader.py (hunk 8 above)
Comment (jrplatin) [reply tpu_commons/models/jax/model_loader.py:57]: Ah good to know, thanks!

File: tests/e2e/benchmarking/mmlu.sh (hunk 7 above)
Comment (gpolovets1) [reply tests/e2e/benchmarking/mmlu.sh]: I just mean why don't you add these extra args if model_name = meta-llama/Llama-3.1-8B-Instruct? Just one less env variable to worry about

Comment (jrplatin) [reply tests/e2e/benchmarking/mmlu.sh]: Ah we want to test both quantized and unquantized models
